### PR TITLE
Prune unused root replace directives

### PR DESCRIPTION
**Pull request by liggitt**
#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:

Makes root go.mod generation idempotent, otherwise we can get fluttering add/remove behavior on alternating runs of `update-vendor.sh` as seen in https://github.com/kubernetes/kubernetes/pull/112015#issuecomment-1244935864

We already pruned unused replace directives from staging `go.mod` files, we forgot to do the same in root `go.mod` files, and had never encountered a combination of dependencies where it mattered.

```release-note
NONE
```

/area code-organization
/sig architecture
/cc @dims @humblec

### CODE REVIEW

File: hack/update-vendor.sh [hunk 1]
```diff
@@ -358,6 +358,13 @@ done
 echo "=== tidying root" >> "${LOG_FILE}"
 go mod tidy >>"${LOG_FILE}" 2>&1
 
+# prune unused pinned non-local replace directives
```
Comment (liggitt) [inline hack/update-vendor.sh:361]: this is the same check as on line 338, with the addition of explicitly keeping local replace directives in the root (`k8s.io/foo => ./staging/src/k8s.io/foo`)

Comment (dims) [reply hack/update-vendor.sh:361]: got it (`select(.New.Path | startswith("./")`)

**Comment by humblec** [comment]
Thanks for the fix @liggitt . lgtm :+1:

**Pull request merged**